pkg/logs/logs.go

```diff
@@ -1,8 +1,164 @@
+//  - [GKE logs best practices](https://cloud.google.com/kubernetes-engine/docs/concepts/about-logs#best_practices)
+//  - [Structured Logging KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/1602-structured-logging/README.md)
+//  - [Examples of using k8s.io/component-base/logs](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/component-base/logs/example),
+//    upon which this code was based.
+
+var (
+	// This is the Agent's logger. For now, it is still a *log.Logger, but we
+	// mean to migrate everything to slog with the klog backend. We avoid using
+	// log.Default because log.Default is already used by the VCert library, and
+	// we need to keep the agent's logger from the VCert's logger to be able to
+	// remove the `vCert: ` prefix from the VCert logs.
+	Log *log.Logger
+
+	// All but the essential logging flags will be hidden to avoid overwhelming
+	// the user. The hidden flags can still be used. For example if a user does
+	// not like the split-stream behavior and a Venafi field engineer can
+	// instruct them to patch --log-json-split-stream=false on to the Deployment
+	// arguments.
+	visibleFlagNames = sets.New[string]("v", "vmodule", "logging-format")
+	// This default logging configuration will be updated with values from the
+	// logging flags, even those that are hidden.
+	configuration = logsapi.NewLoggingConfiguration()
+	// Logging features will be added to this feature gate, but the
+	// feature-gates flag will be hidden from the user.
+	features = featuregate.NewFeatureGate()
 )
 
-var Log = log.New(os.Stderr, "", log.LstdFlags)
+func init() {
+	runtime.Must(logsapi.AddFeatureGates(features))
+	// Turn on ALPHA options to enable the split-stream logging options.
+	runtime.Must(features.OverrideDefault(logsapi.LoggingAlphaOptions, true))
+}
+
+// AddFlags adds log related flags to the supplied flag set.
+//
+// The default logging format is changed to JSON. The default in Kubernetes
+// component base is "text", for backwards compatibility, but that is not a
+// concern for venafi-kubernetes-agent.
+// The split-stream options are enabled by default, so that errors are logged to
+// stderr and info to stdout, allowing cloud logging systems to assign an
+// severity INFO or ERROR to the messages.
+func AddFlags(fs *pflag.FlagSet) {
+	var tfs pflag.FlagSet
+	logsapi.AddFlags(configuration, &tfs)
+	features.AddFlag(&tfs)
+	tfs.VisitAll(func(f *pflag.Flag) {
+		if !visibleFlagNames.Has(f.Name) {
+			tfs.MarkHidden(f.Name)
+		}
+
+		// The original usage string includes details about how
+		// JSON logging is only available when BETA logging features are
+		// enabled, but that's not relevant here because the feature is enabled
+		// by default.
+		if f.Name == "logging-format" {
+			f.Usage = `Sets the log format. Permitted formats: "json", "text".`
+		}
+		if f.Name == "log-text-split-stream" {
```

Comment:
Here's my current understanding:
1. I enabled `log-json-split-stream` by default to solve the problem that Google Logs Explorer otherwise can't distinguish between JSON formatted errors and info messages.
    > Severities: By default, logs written to the standard output are on the INFO level and logs written to the standard error are on the ERROR level. Structured logs can include a severity field, which defines the log's severity.
    > -- https://cloud.google.com/kubernetes-engine/docs/concepts/about-logs#best_practices
3. `log-text-split-stream` is not technically necessary to solve that problem because Google Logs Explorer seems to parse the `E`, `W`, and `I` prefix of the Kubernetes text format. So yes, I could remove this change.
4. Kubernetes introduced the split-stream options to solve performance problems. Info and warn messages can be buffered and flushed periodically while errors are flushed immediately.
    * https://github.com/kubernetes/kubernetes/pull/104873
    * https://github.com/kubernetes/enhancements/pull/2912#discussion_r703346009
6. By default logs are not buffered and are flushed to stdout and stderr immediately, so there shouldn't be any out-of-order log messages.
7. If log buffering is enabled, then I *think* that the buffered info logs will be flushed whenever an error is logged, so there should be any out-of-order log messages in that case either.
8. Kubernetes plan to make split-stream JSON logging the default for the Kubernetes components.
   > Ability to route logs based on type/verbosity will be replaced with default splitting info and errors logs to stdout and stderr. We will make this change only in alternative logging formats (like JSON) as we don't want to introduce breaking change in default configuration. Splitting stream will allow treating info and errors with different priorities. It will unblock efforts like https://github.com/kubernetes/klog/issues/209 to make info logs non-blocking.
   > 
   > -- https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components#logging-defaults